pallets/confidential-asset/src/lib.rs

```diff
@@ -370,20 +404,247 @@ decl_module! {
+            Self::deposit_event(RawEvent::ConfidentialAccountNewCheckpoint(
+                owner_did,
+                account_id,
+                current_balance
+            ));
+
+            Ok(())
+        }
     }
 }
 
 impl<T: Trait> Module<T> {
-    pub fn update_mercat_account_balance(
+    /// Reset `IncomingBalance` and `FailedOutgoingBalance` accumulators, by doing this
+    /// we are no longer keeping track of the old/settled transactions.
+    fn reset_mercat_pending_state(owner_did: &IdentityId, account_id: &MercatAccountId) {
+        if <IncomingBalance>::contains_key(owner_did, account_id.clone()) {
+            <IncomingBalance>::insert(
+                owner_did,
+                account_id,
+                EncryptedBalanceWrapper::from(EncryptedAmount::default().encode()),
+            );
+        }
+
+        if <FailedOutgoingBalance>::contains_key(owner_did, account_id.clone()) {
+            <FailedOutgoingBalance>::insert(
+                owner_did,
+                account_id,
+                EncryptedBalanceWrapper::from(EncryptedAmount::default().encode()),
+            );
+        }
+    }
+
+    /// Add the `amount` to the mercat account balance, and update the `IncomingBalance` accumulator.
+    pub fn mercat_account_deposit_amount(
         owner_did: &IdentityId,
         account_id: &MercatAccountId,
-        new_encrypted_balance: EncryptedAmount,
-    ) {
+        amount: EncryptedAmount,
+    ) -> DispatchResult {
+        let current_balance =
+            Self::mercat_account_balance(owner_did, account_id.clone()).to_mercat::<T>()?;
+        <MercatAccountBalance>::insert(
+            owner_did,
+            account_id,
+            EncryptedBalanceWrapper::from((current_balance + amount).encode()),
+        );
+
+        // Update the incoming balance accumulator.
+        Ok(Self::add_incoming_balance(owner_did, account_id, amount)?)
+    }
+
+    /// Subtract the `amount` from the mercat account balance, and update the `PendingOutgoingBalance` accumulator.
+    pub fn mercat_account_withdraw_amount(
+        owner_did: &IdentityId,
+        account_id: &MercatAccountId,
+        amount: EncryptedAmount,
+    ) -> DispatchResult {
+        let current_balance =
+            Self::mercat_account_balance(owner_did, account_id.clone()).to_mercat::<T>()?;
         <MercatAccountBalance>::insert(
             owner_did,
             account_id,
-            EncryptedBalanceWrapper::from(new_encrypted_balance.encode()),
+            EncryptedBalanceWrapper::from((current_balance - amount).encode()),
+        );
+
+        // Update the pending balance accumulator.
+        let pending_balance =
+            Self::pending_outgoing_balance(owner_did, account_id.clone()).to_mercat::<T>()?;
+
+        <PendingOutgoingBalance>::insert(
+            owner_did,
+            account_id,
+            EncryptedBalanceWrapper::from((pending_balance - amount).encode()),
```

Comment:
Subtraction can underflow. Please use `checked_sub` like in the balances module and return an error on underflow.